### ISSUE TITLE
Issue loading bgzipped indexed bedGraph files

### ISSUE THREAD

**Issue by vidboda** [igvteam/igv.js#1654]
Hi,

I'm trying to load small bgzipped indexed by tabix bedGraph files (about 100K, 20,000bp) which I use in my webapp to store spliceai scores per DNA variants. They are currently in bedGraph format but as the number of variants is growing it currently represents ~100Go of data. 

That's why I'd like to use them compressed.

I saw this [issue](https://github.com/igvteam/igv.js/issues/320) and understood that this type of files could be handled by the amazing igv.js.

However I've got some troubles. When trying to load one track like this:

```
              {
                name: 'Test gz',                
                type: 'wig',
                format: 'bedgraph',
                min: -1,
                max: 1,
                url: static_path + 'resources/spliceai/variants/100658.bedGraph.gz',
                indexURL: static_path + 'resources/spliceai/variants/100658.bedGraph.tbi',
                label: 'Test gz'                
              },
```

in firefox (macos), I'm getting an "invalid block type" error coming from bgzf.js, from this block of code (lines 34-61), (and no display at all in igv.js):

```
   while (ptr < lim) {
        try {
            const ba = ArrayBuffer.isView(data) ? data : new Uint8Array(data, ptr, 18);
            const xlen = (ba[11] << 8) | (ba[10]);
            const flg = ba[3];
            const fextra = flg & FEXTRA;
            const si1 = ba[12];
            const si2 = ba[13];
            const slen = (ba[15] << 8) | (ba[14]);
            const bsize = ((ba[17] << 8) | (ba[16])) + 1;
            const start = 12 + xlen + ptr;    // Start of CDATA
            const bytesLeft = data.byteLength - start;
            const cDataSize = bsize - xlen - 19;
            if (bytesLeft < cDataSize || cDataSize <= 0) break;

            const a = new Uint8Array(data, start, cDataSize);
            const unc = inflateRaw(a);

            // const inflate = new Zlib.RawInflate(a);
            // const unc = inflate.decompress();

            ptr += (cDataSize - 1) + 26; //inflate.ip + 26
            totalSize += unc.byteLength;
            oBlockList.push(unc);
        } catch (e) {
            console.error(e)
            break;
        }
```

That I am not able to debug.

Some additional precisions:

- the file I'm attemting to load was successfully indexed by tabix and I'm able to query it using the tabix command line.
- I'm using igv.js v2.15.8
- I'm able with the same config to load the bedGraph from the 2017 issue:

```
              {
                'url':'https://s3-us-west-2.amazonaws.com/ilmn.igv-test/test.bedgraph.gz',
                'indexURL':'https://s3-us-west-2.amazonaws.com/ilmn.igv-test/test.bedgraph.gz.tbi',
                'type':'wig',
                'format':'bedgraph',
                'name':'test-bg-tbi-aswig'
              },
```

- here's a sample of the bedGraph file I'm trying to load:

```
chr1    216412303       216412304       0.00000892                                                                          
chr1    216412304       216412305       0.00000238
chr1    216412305       216412306       0.00000081
chr1    216412306       216412307       0.00000370
chr1    216412307       216412308       0.00000234
chr1    216412308       216412309       0.00002424
chr1    216412309       216412310       0.00000269
chr1    216412310       216412311       0.00000193
chr1    216412311       216412312       0.00002411
chr1    216412312       216412313       0.00000216
chr1    216412313       216412314       -0.0002605
chr1    216412314       216412315       0.00000402
chr1    216412315       216412316       0.00000113
chr1    216412316       216412317       0.00000212                                                                          
chr1    216412317       216412318       0.00000397                                                                                                                                                                                                      
chr1    216412318       216412319       0.00000078                                                                                                                                                                                                      
chr1    216412319       216412320       0.00001720                                                                          
chr1    216412320       216412321       0.00000427                                                                          
chr1    216412321       216412322       -1.583e-05                                                                                                                                                                                                      
chr1    216412322       216412323       0.00000081                                                                          
chr1    216412323       216412324       0.00000590                                                                                                                                                                                                      
chr1    216412324       216412325       0.00000145                                                                          
chr1    216412325       216412326       0.00000089                                                                                                                                                                                                      
chr1    216412326       216412327       0.00000121                                                                          
chr1    216412327       216412328       0.00000389                                                                                                                                                                                                      
chr1    216412328       216412329       0.00000063               
```

I also wondered if this was maybe linked to the type of notation '-1.583e-05' that triggers the error, but tried a cleaned file and got the same result.

Thanks in advance for any help you could provide!

**Comment by jrobinso** [comment]
That should work,  can you share the a test file (actually the pair, both gz and tbi)?

Also, coul you try this test case,  but download the reference files and put them on your data server (whatever `static_path` refers to).  This will help us eliminate server side issues.

```
        {
            genome: "hg38",
            locus: "myc",
            tracks: [
                {
                    url: "https://s3.amazonaws.com/igv.org.test/data/bedgraph/ENCFF439GUL.bedgraph.gz",
                    indexURL: "https://s3.amazonaws.com/igv.org.test/data/bedgraph/ENCFF439GUL.bedgraph.gz.tbi",
                    name: "RNA-Seq",
                    type: "wig",
                    format: "bedgraph",
                    visibilityWindow: 10000000
                },
            ]
        }
```

**Comment by beboche** [comment]
Thanks for the answer.

You are right, I may be facing a server side issue, as when I download your dataset the app triggers the same "invalid block type" error. I'm currently running these tests on a Flask app using the embedded flask server (which however handles correctly gzipped fasta files for reference, e.g.:

```
reference: {
              id: "hg38",
              name: "Human (GRCh38/hg38)",
              fastaURL: static_path + "resources/genome/hg38.fa.gz",
              indexURL: static_path + "resources/genome/hg38.fa.gz.fai",
              compressedIndexURL: static_path + "resources/genome/hg38.fa.gz.gzi",
              cytobandURL: static_path + "resources/genome/cytoBandIdeo.txt.gz"
            },
```

I'll try to dig this way, and try with a version that runs with apache instead of the flask server.

In addition, here is a temporary [link](https://filesender.renater.fr/?s=download&token=4c60139f-7f69-44c3-8505-88e736ddf117) to the a test dataset.

**Comment by jrobinso** [comment]
I can't reproduce the problem with your test file,  my configuration is below.    However I did notice a typo in your example, the url to the tabix index file  is `....bedGraph.tbi'` rather than `....bedGraph.gz.tbi'`.    This should not cause the symptom you report, but could depending on how the server responds for a file not found error.


```
<script type="module">

    import igv from  "https://cdn.jsdelivr.net/npm/igv@2.15.8/dist/igv.esm.min.js"

    let browser
    const options =
        {
            genome: "hg38",
            locus: "chr1:216,412,242-216,412,281",
            tracks: [
                {
                    name: 'Test gz',
                    type: 'wig',
                    format: 'bedgraph',
                    min: -1,
                    max: 1,
                    url: '100658.bedGraph.gz',
                    //indexed: false,
                    indexURL: '100658.bedGraph.gz.tbi',
                    label: 'Test gz'
                }]
        }

    igv.createBrowser(document.getElementById('igvDiv'), options)
        .then(function (b) {
            browser = b
        })


</script>

**Comment by jrobinso** [comment]
@beboche  The browser's developer tools network tab can be informative.   The error you are seeing indicates the javascript is not seeing the bytes it expects.    Here's a screenshot from the example above with Firefox web developer tools on and the network tab selected.     Not especially the "Size"  column .   If you run the same example you should get exactly the same result.   If you don't that will be a hint.  

Selecting the first request to "bedGraph.gz" I get the second screenshot.    Note in the response headers the content-length is 26 and content range is 0-25/428.   If you're values are different again that is a clue.

Since you are seeing this from 2 different servers we can probably rule them out.   Requests and responses might be going through something else,  a proxy or virus protection for example, that is manipulating the results.  Sometimes proxies can strip request headers, such as "Range", for example.   That is really unlikely in this case.  In any event check the responses using the example posted above with  developer tools and compare with these screenshots.

<img width="1581" alt="Screen Shot 2023-06-27 at 9 46 21 AM" src="https://github.com/igvteam/igv.js/assets/933148/bd132167-53ef-4cc3-a930-518fb3283d99">
<img width="1646" alt="Screen Shot 2023-06-27 at 9 44 03 AM" src="https://github.com/igvteam/igv.js/assets/933148/439a3810-0165-4082-9310-f967dc489d2d">

**Comment by beboche** [comment]
Hi @jrobinso ,

thanks for the follow-up.

ok using your test code from June 26 through the apache server, everything looks now weirdly fine (I just modified the bedgraph file to load a complete file, so the range is now 0-25/116302):

<img width="1689" alt="image" src="https://github.com/igvteam/igv.js/assets/29146185/1ea68d47-d78d-4ee1-9437-c0b9f7138898">

And the bedgraph shows up.
That's already great news.

Now, I'm testing this embedded in my Flask app. Reminder, in this app, the igv.js already successfully displays generic bedGraph files, and also the gzipped genome (see above). So no CORS issue, and then the browser does not complain about it.

Weirdly again, I no longer see the bgzf.js error. Not sure what happens, the code being the same. Anyway, in this case, the network metrics look good:

<img width="1704" alt="image" src="https://github.com/igvteam/igv.js/assets/29146185/403b3e1b-9380-45f2-93b2-c6c71cfbd6e1">

But, the bars do not appear on the track (no orange baseline as in the first screencapture).

If I zoom out the test page and the app page, it will be more clear:

test page:

<img width="960" alt="image" src="https://github.com/igvteam/igv.js/assets/29146185/21a1d20f-2964-459a-bb33-4a655dbd5fcb">

app page with empty track:

<img width="1417" alt="image" src="https://github.com/igvteam/igv.js/assets/29146185/f19c3afb-9f06-4fe9-869b-12aead844450">

I need to mention 2 more things:

the compress bedgraph file in both example has this header:
```
track name="ALT allele (MobiDetails ID: 100658)" type=bedGraph description="spliceAI predictions for variant 100658 in MobiDetails    acceptor_sites = positive_values       donor_sites = negative_values" visibility=full windowingFunction=maximum color=200,100,0 altColor=0,100,200 priority=20 autoScale=off viewLimits=-1:1 darkerLabels=on
```

which among other things sets the view limits [-1;1]. We can see in the second example that the limits are correctly set, so I understand that the file is read by igv.js. Just the bars do not appear. I can provide you this second example file, if you like. This very same file is correctly loaded in the test page.

Second, the code I'm using in the app example (from a .js file, simplified here for this test, i.e. no other tracks):

```
var options =
          {
            showNavigation: true,
            showRuler: true,
            showCenterGuide: true,
            showCursorTrackingGuide: true,
            locus: 'chr1:216,421,925-216,422,549',
            reference: {
              id: "hg38",
              name: "Human (GRCh38/hg38)",
              fastaURL: static_path + "resources/genome/hg38.fa.gz",
              indexURL: static_path + "resources/genome/hg38.fa.gz.fai",
              compressedIndexURL: static_path + "resources/genome/hg38.fa.gz.gzi",
              cytobandURL: static_path + "resources/genome/cytoBandIdeo.txt.gz"
            },
            tracks: [
              {
                name: 'SpliceAI MT NM_206933.2',
                type: 'wig',
                format: 'bedgraph',
                url: static_path + 'resources/spliceai/variants/100658.bedGraph.gz',
                indexURL: static_path + 'resources/spliceai/variants/100658.bedGraph.gz.tbi',
                label: 'SpliceAI raw scores for NM_206933.2' ,
              },
              {
                name: 'MANE transcripts',
                url: static_path + 'resources/mane/MANE.GRCh38.v1.0.refseq.bb',
                indexed: false,
                label: 'MANE transcripts',
              },
              {
                name: 'Refseq Genes',
                url: static_path + 'resources/genome/refGene.txt.gz',
                order: 1000000,
                indexed: false
              }
            ]
          };
        igv.createBrowser(document.getElementById('igv_div'), options)
          .then(function (browser) {
              igv.browser = browser;
              if (spliceaivisual_response == 'full') {
                // load full transcript track
                add_full_gene_track(static_path);
              }
          });
```

To summarize:

loading a smple web page with the code from June 26 works fine through Apache, but not when embedded in a flask app (with a different way of loading igv.js, and no bgzf.js error anymore in the console, for an unknown reason).

I'll now try to load the code that works through apache from the flask app.

**Comment by beboche** [comment]
so, no difference if I load a simple page from the flask app:

- no error in Firefox console
- the network tab displays correct values (the same as above)
- the track is correctly centered between -1 and +1 but remains empty.

However I noticed this in the firefox console:

<img width="1688" alt="image" src="https://github.com/igvteam/igv.js/assets/29146185/4d06faa6-2847-4818-970c-b841fc938ee3">

which means in english something like: "colour expected, but 'undefined' found", from DOM.

I then loaded the same page with brave browser:

<img width="1633" alt="image" src="https://github.com/igvteam/igv.js/assets/29146185/fdadb791-5df9-47a9-ba08-6dea45152085">

And the bedgraph loaded successfully. Safari is fine, too. Then Firefox seems to have a problem to properly load the track, when coming form the flask app. Not sure if it's still igv.js related, but quite annoying.

**Comment by jrobinso** [comment]
Is this a correct summary,  everything works in all combinations of server + browser EXCEPT flask + firefox?   Are there any errors in the javascript console?   Is the network showing exactly the same responses with flask + firefox as any-other-server + firefox?

I will try to reproduce this if you can zip up a complete example, by that I mean html + data files.   I can edit the url paths to the data files when I set it up here.   Also exactly which version of flask + anything special I need to know about configuring it.

**Comment by beboche** [comment]
Correct.

I'l try to prepare all you need to reproduce.

**Comment by beboche** [comment]
I tried to set up the simplest flask app to help reproduce. The archive is available below:

[test_igv.tgz](https://github.com/igvteam/igv.js/files/11897523/test_igv.tgz)

To run it you need python together with the Flask module. Below my versions: python 3.8 and flask 2.3.2

```
$ python3.8 --version
Python 3.8.9
$ python3.8 -c "import flask; print(flask.__version__)"
2.3.2
```

The most proper way would be to use a virtualenv or maybe better a conda env, then, from the test_igv folder (location on the server does not matter):

```
$ python3.8 -m flask --debug run --host=0.0.0.0  --port=5000
```
Everything here is served via the flask embedded server (which gave me the same results as apache on the previous tests)
You also need to open port 5000/tcp or modify the port of the startup command line. --host=0.0.0.0 will allow you to see the page in your browser using the IP address:port of your server, e.g. http://your.ip:5000. http://127.0.0.1:5000 should also work if you run it locally.

Of note the `app.config['USE_X_SENDFILE'] = False`line in the test_igv.py file which seems mandatory to properly send a file using the flask server (different from a production server).

Anyway, with this app the results I get are:

- firefox:

error in the console: 

<img width="1691" alt="image" src="https://github.com/igvteam/igv.js/assets/29146185/940bd671-c09e-4c8e-9970-02f1d3ce3f58">

network:

<img width="1690" alt="image" src="https://github.com/igvteam/igv.js/assets/29146185/62058443-c5dc-4b13-93b9-7c77c7e035b6">

- brave:

error in the console:

<img width="2046" alt="image" src="https://github.com/igvteam/igv.js/assets/29146185/6d02ea37-05d4-4d4e-adb9-02ce74d60060">

network:

<img width="842" alt="image" src="https://github.com/igvteam/igv.js/assets/29146185/11e340f0-a727-41d1-8b66-c35991373c30">


- safari:

fine (I zoomed out to show):

<img width="1430" alt="image" src="https://github.com/igvteam/igv.js/assets/29146185/e538ebe4-efb0-48e5-9882-d85432e29766">

If I set app.config['USE_X_SENDFILE'] to True, Safari also fails.

**Comment by jrobinso** [comment]
I also get the error in Chrome.

There is something not right about this flask app.  I will try to debug it but python and flask are not my strong points.    I do know many sites that use flask with igv.js,  in fact it might be the most used server,  but unfortunately I'm not an expert in it.

BTW "USE_X_SENDFILE" should default to false.   What does "app.config['SECRET_KEY'] do?

**Comment by jrobinso** [comment]
I think I found the cause, but not the solution yet.   I think the problem is the server is inserting the response header `content-encoding: gzip`.    This tells the browser to gunzip the data before javascript even sees it.   This will not work with bgzipped files.   We had a similar problem once with some Apache servers adding this header for BAM files.

I do not understand how Safari is working, that is a mystery, unless it ignores content-encoding if content-type is set to octet-stream.

If you can figure out a way to stop Flask from setting content-encoding I think it will work.   I will return to this later today.

**Comment by jrobinso** [comment]
Its possible Flask is gzipping this file, and thus the header,  if this is the case the task to to turn off compression for already gzipped (in this case block gzipped) files.

**Comment by beboche** [comment]
thanks for the feedback. 

> BTW "USE_X_SENDFILE" should default to false. What does "app.config['SECRET_KEY'] do?

just setting a secret key for the app, which is important on production environment. here I just set something different than `None`.

It seems to be used to sign cookies (https://stackoverflow.com/questions/22463939/demystify-flask-app-secret-key).

**Comment by jrobinso** [comment]
Well I've spent a bit more time on this but don't have a solution.    Here's a faster way to debug though with curl.  If `Content-Encoding: gzip` is present in the response headers (see below) its not going to work.  Again I don't know why it works on Safari, that is the real suprise,  but perhaps Safari ignores Content-Encoding if Content-Type is octet-stream.

The `Content-Disposition` header is strange also, and might well cause problems.   

```
curl -i  http://127.0.0.1:5000/static/100658.bedGraph.gz
HTTP/1.1 200 OK
Server: Werkzeug/2.3.6 Python/3.8.7
Date: Wed, 28 Jun 2023 19:28:15 GMT
Content-Encoding: gzip
Content-Disposition: inline; filename=100658.bedGraph.gz
Content-Type: application/octet-stream
Content-Length: 116302
Last-Modified: Wed, 28 Jun 2023 15:13:00 GMT
Cache-Control: no-cache
ETag: "1687965180.0-116302-1315251226"
Date: Wed, 28 Jun 2023 19:28:15 GMT
Connection: close
```

**Comment by beboche** [comment]
ok thanks for the curl hint, I'll make some google tomorrow on those headers

**Comment by jrobinso** [comment]
I think I got it.  Try using `@app.after_request` to remove the `content-encoding` header as shown below.  This works for me, and I can run the example in the header.    This is not a perfect solution,  if you set up compression and Flask compressed a plain text file, for example, the content-encoding header would be needed to reverse the compression in the client.   But that's not what we are doing here.    

So I think this is a solution, but it would need refined if you want the flask server to do gzip compression for you.

I think this is a Flask bug,  it did not compress the file so it should not be assuming the client wants it to be uncompressed.

```
#!/usr/local/bin/python3.8
from flask import Flask, render_template

app = Flask(__name__, static_folder='static')
app.config['USE_X_SENDFILE'] = False
app.config['SECRET_KEY'] = 'test_igv'
print(app.config)

@app.route('/')
def test_igv():
    return render_template('test_igv.html')

@app.after_request
def remove_header(response):
    del response.headers['content-encoding']
    return response
```

**Comment by jrobinso** [comment]
For completeness I suggest removing `content-disposition` as well.  Although it doesn't seem to be causing problems, it can be doing anything useful.

```
@app.after_request
def remove_header(response):
    del response.headers['content-encoding']
    del response.headers['content-disposition']
    return response
```

**Comment by beboche** [comment]
Brilliant! I'll test this and see how to introduce it in a more complex app (i.e. not breaking the rest of the app!).

Thanks again

**Comment by beboche** [comment]
Well it looks all good so far. I need to do more testings but this looks solved. Many thanks again!!

**Comment by jrobinso** [comment]
The best solution is to figure out why Flask is inserting this header in the first place.   The presence of the 'content-disposition' header might be a hint, it indicates the content is to be displayed inline on the page.    Whatever is causing that might be triggering the 'content-encoding: gzip' header.   If you find out what is causing this and how to disable it please post here.

**Comment by beboche** [comment]
right, I'll do that if I fond something.

**Comment by jrobinso** [comment]
Also, you might try determining if the response is for a content range, by examining the response object, and only remove the headers if it is.   For example you could check the status code, it should be 206 for content range requests.  I can't think of any situation where a request would be made for a specific range of bytes that did not expect to receive exactly that range of bytes, unmodified.   Also, gunzipping by the browser on a slice of a gzipped file is almost never going to work.   By doing this check you could possibly remove the header when its harmful,  but leave it in other cases.     

Better to solve the root problem but that might be difficult.

**Comment by jrobinso** [comment]
For example, like this

```
@app.after_request
def remove_header(response):
    if response.status_code == 206:
        del response.headers['content-encoding']
        del response.headers['content-disposition']
    return response
```

**Comment by beboche** [comment]
That's precisely the code I'm testing currently ;-)

**Comment by beboche** [comment]
Hi @jrobinso,

Thanks again for solving this flask issue. As I also serve webpages where the bedgraphs can be fully loaded (HTTP return code 200), I finally went up with the current fix which looks to work quite fine:

in the create_app() function of my flask app, I simply added the following code, taking advantage of the request object:

You need to import the request object first:

```
from flask import request
```

Then you can add:

```
@app.after_request
    def remove_header(response):
        # remove headers only for bedgraph.gz files
        if re.search('bedGraph.gz$', request.url):
            del response.headers['content-encoding']
            del response.headers['content-disposition']
        return response
```

Works with the flask server and on my apache test server also, without any modif on the apache side.

Not yet tested in production.

**Comment by jrobinso** [comment]
Thanks for the update, keep us informed.   This should prove useful for others who find this issue. 

Do you server "bam" or "cram" files?  You might have the same issue with these files.   They are also bgzipped.

**Comment by beboche** [comment]
No I don't serve BAMs or CRAMs with this app - I do with another, but it's older and perl based and I have no issue so far.

However I'm using compressed (bgzipped) reference with this code, for months:

```
reference: {
  id: "hg38",
  name: "Human (GRCh38/hg38)",
  fastaURL: static_path + "resources/genome/hg38.fa.gz",
  indexURL: static_path + "resources/genome/hg38.fa.gz.fai",
  compressedIndexURL: static_path + "resources/genome/hg38.fa.gz.gzi",
  cytobandURL: static_path + "resources/genome/cytoBandIdeo.txt.gz"
},
```

Weirdly, since this issue, the genome was not loaded anymore (cytobands were ok). I also load a refseq gene track (refGene.txt.gz, classical gzip, I suppose) and a MANE track (MANE.GRCh38.v1.0.refseq.bb) which are ok.

I adapted the above code in my flask app for the genome and it seems to work on my test server:

```
@app.after_request
def remove_header(response):
    # remove header for bedgraph.gz and genome.gz files
    if re.search('(bedGraph|fa).gz$', request.url):
        del response.headers['content-encoding']
        del response.headers['content-disposition']
    return response
```

I am not sure why the genome is not loading correctly anymore.

The current production server uses the old code with uncompressed bedgraphs and compressed genome and no header removal and everything's fine to me (e.g. https://mobidetails.iurc.montp.inserm.fr/MD/api/variant/97699/browser/#splicing, scrool down a bit to find the igv.js browser, you should see the gnome correctly loaded).

However, if I do not remove the headers for the genome on my dev server using compressed bedgraph, the genome does not load correctly:

<img width="1769" alt="image" src="https://github.com/igvteam/igv.js/assets/29146185/408b4b83-d31a-4bce-9392-cdd215b71d10">


This is no big deal and if removing headers works correctly with the genome from now on, I'm fine with it, but I just don't understand what's going on between the 2 versions.

**Comment by jrobinso** [comment]
I don't know what's going on with FLASK either but if it sets "Content-Encoding: gzip" on any bgzipped file that is not loaded in its entireity, and that would include your compressed fasta, its not going to work.    That header causes the browser to gunzip the content before IGV even sees it.    BGZipped files have to be gunzipped in a particular way,  the browser will not do it correctly, and even if it did its not what IGV expects.

The reason your regular gzipped file works, even with the Content-Encoding header,  is because the content is gunzipped correctly, and IGV gets the contents of the file which it is happy with.

Flask should definitely not be adding that header since it did not compress the content,  if there is an issue tracker for Flask you might open an issue for this.   It did not used to do that,  some years ago we served content with Flask to IGV and we did not encounter this issue.   

If that header is present in a file that is known to be BGZipped it should always be removed.

**Comment by jrobinso** [comment]
The production server works because Content-Encoding is not added (see screenshot).   Is this a Flask server also? 

<img width="631" alt="Screen Shot 2023-07-05 at 1 45 01 PM" src="https://github.com/igvteam/igv.js/assets/933148/04ed1e77-1ac3-4bd7-b648-7570a8d061a0">

**Comment by beboche** [comment]
> The production server works because Content-Encoding is not added (see screenshot). Is this a Flask server also?

well, a Flask app with apache server, but the flask version is older on the production server, which may explain the discrepancy
